### PR TITLE
feat(UVE): Bring back inline editing for block editor fields in Headless

### LINKED ISSUE #30607: UVE: Bring back inline editing for block editor fields in Headless
### Parent
#27546 

### Task description
As a web content editor using dotCMS, I want to have the inline editing Block Editor in headless, so that I can easily make quick edits to content without switching to the code view.

**Wait for #28198**

https://github.com/dotCMS/core/assets/751424/1031f189-37ee-4aba-b2b1-eb8030fd46f0

### Acceptance Criteria
1. The inline editing feature is working with headless implementations
2. Users can seamlessly edit content inline without any issues.
3. The functionality works the same as it did in the old editor.
4. Users are not forced to update their code when using the inline editing feature.

### External Links
- Slack conversation: #dotCMS-development
- Support ticket: #12345

### Assumptions & Initiation Needs
- The new editor interface has the capability to support the inline editing feature.
- Users are familiar with how the inline editing feature worked in the old editor.
- A thorough testing phase will be conducted to ensure the feature works as expected.

### Quality Assurance Notes & Workarounds
- QA team should test the feature extensively to ensure it works seamlessly.
- Any issues or bugs encountered during testing should be documented and addressed before release.

### PR DESCRIPTION
**Pull request by rjvelazco**
This pull request introduces significant changes to the inline editing functionality within the `core-web` library, primarily focusing on the block editor and WYSIWYG editor integration. The updates include modifications to the event handling, data structures, and messaging between components. Below are the most important changes grouped by theme:

### Inline Editing Event Handling:

* Introduced a new `#handleInlineEditingEvent` method in `EditEmaEditorComponent` to handle different types of inline editing events, such as `BLOCK_EDITOR` and `WYSIWYG`. This method checks for enterprise licenses and handles the event accordingly.
* Updated the `CLIENT_ACTIONS` enum to remove `INIT_BLOCK_EDITOR_INLINE_EDITING` and added `INIT_INLINE_EDITING` to streamline the event handling process.

### Data Structure Updates:

* Modified the `InlineEventData` interface to replace `blockEditorContent` with a structured `content` object in `dot-block-editor-sidebar.component.ts`. This change ensures that the content data is more organized and easier to manage.
* Updated the `BlockEditorData` interface to include a `JSONContent` type for the `content` field, enhancing the data consistency and type safety.

### Messaging and Communication:

* Refactored the `initInlineEditing` function to send messages to the editor with the appropriate payload for both `BLOCK_EDITOR` and `WYSIWYG` types. This change standardizes the communication between the client and the editor.
* Added tests for the `initInlineEditing` function to ensure it sends the correct messages based on the provided type and data.

### Component and Service Adjustments:

* Removed the `#getJsonContent` method from `DotBlockEditorSidebarComponent` as it is no longer needed with the updated `content` handling.
* Updated the `InlineEditService` to use the new `CLIENT_ACTIONS.INIT_INLINE_EDITING` action, ensuring consistent event handling across the application.

### Test Updates:

* Modified existing tests and added new ones to cover the changes in event handling, data structures, and messaging. These updates ensure that the new functionality is thoroughly tested and reliable. [[1]](diffhunk://#diff-91349cd92f693c5f3ef517eb5ab5e553167c142d330bce29621b892583b2c177L14-R41) [[2]](diffhunk://#diff-bae6613453ca5e03886194e908b1dbed1ffe067808be88cd178dbc8defa024ccR128-R164)

### Headless

https://github.com/user-attachments/assets/4b8dc731-f805-46f2-8910-05485112b174

### Traditional

https://github.com/user-attachments/assets/66a8967a-438c-4b45-9c78-f052326ac872

### WYSIWYG

https://github.com/user-attachments/assets/c4d2087c-932b-43ed-ba5f-1b6c4bed0fc5

This PR fixes: #30607

### CODE REVIEW

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 1]
```diff
@@ -1365,4 +1352,44 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
+            case 'WYSIWYG':
+                this.inlineEditingService.initEditor();
+                this.uveStore.setEditorState(EDITOR_STATE.INLINE_EDITING);
+                break;
+
+            default:
+                console.warn('Unknown block editor type', type);
+
+                break;
+        }
```
Comment (zJaaal) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:1393]: Awesome

**Review by zJaaal — DISMISSED** [review]

File: core-web/libs/sdk/react/src/lib/components/BlockEditorRenderer/BlockEditorRenderer.tsx [hunk 2]
```diff
@@ -19,16 +38,57 @@ export interface BlockEditorRendererProps {
...
+        // but we need to check them again to avoid runtime errors in Pure JavaScript
+        if (!contentlet || !fieldName) {
+            console.error('contentlet and fieldName are required to enable inline editing');
+
+            return;
+        }
+
+        const { inode, languageId: language, contentType } = contentlet;
+        // `ContentNode` lives on `@dotcms/react` that's why we can use it in `@dotcms/client`
+        // We need to move interfaces to external lib
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/sdk/react/src/lib/components/BlockEditorRenderer/BlockEditorRenderer.tsx:72]: Let's do it. Types should be included in a type library. It is not good practice to use `@dotcms/client/react/angular` to share types among libraries.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts (hunk 1 above)
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:1393]: There are some things that need review, such as `EDITOR_STATE.INLINE_EDITING` and the `inline-editing` service. I believe they are not needed. It's going to be handled in a separate ticket

File: core-web/libs/sdk/client/src/index.ts [hunk 3]
```diff
@@ -30,5 +36,9 @@ export {
     initEditor,
     updateNavigation,
     destroyEditor,
-    ClientConfig
+    ClientConfig,
+    initInlineEditing,
+    InlineEditEventData,
+    InlineEditorData,
+    INLINE_EDITING_EVENT_KEY
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/sdk/client/src/index.ts:43]: I would like to start by creating the dotCMS type before being stable

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts (hunk 1 above)
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:1393]: Also: I believe this and `CLIENT_ACTIONS` could be a service, this components it over 1000 lines, that's crazy

Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:1393]: Not sure if it should be a service, but maybe we can refactor this to reduce, lines in the component

**Review by zJaaal — APPROVED** [review]

**Review by KevinDavilaDotCMS — APPROVED** [review]

**Pull request merged**